File: src/main/python/litcoder/module-2/lab-1/SimpleTextEditor.py
import re
from collections import deque


class CustomStack:
    def __init__(self):
        self.__editor: deque[str] = deque()
        self.__temp: deque[str] = deque()

    def insert(self, val: str) -> None:
        self.__editor.append(val)

    def delete(self, count: int) -> None:
        while True:
            len_top_editor = len(self.__editor[-1])
            if count - len_top_editor <= 0:
                break

            count -= len_top_editor
            self.__temp.append(self.__editor.pop())

            # check for deleting entire string
            if len(self.__editor[-1]) == count:
                self.__editor.pop()
            else:
                self.__editor[-1] = self.__editor[-1][0 : count - 1]
            self.__cleanup()

    def get(self, index: int) -> str:
        for text in self.__editor:
            if index > len(text):
                index -= len(text)
                continue
            return text[index - 1]

        return ""

    def undo(self) -> None:
        self.__editor.pop()

    def __cleanup(self) -> None:
        while self.__temp:
            self.__editor.append(self.__temp.pop())


editor = CustomStack()
commands = re.split(r"[ ,]", input())
i = 0
while i < len(commands):
    increment = 2
    if commands[i] == "1":
        editor.insert(commands[i + 1])
    elif commands[i] == "2":
        editor.delete(int(commands[i + 1]))
    elif commands[i] == "3":
        print(editor.get(int(commands[i + 1])))
    else:
        editor.undo()
        increment = 1

    i += increment

File: codes/college/litcoder/module-2/lab-1/SimpleTextEditor.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Scanner;

public class SimpleTextEditor {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        CustomStack editor = new CustomStack();

        String[] commands = sc.nextLine().split("[ ,]");
        int i = 0;
        while (i < commands.length) {
            int increment = 2;
            if (commands[i].equals("1")) {
                editor.insert(commands[i + 1]);
            } else if (commands[i].equals("2")) {
                editor.delete(Integer.parseInt(commands[i + 1]));
            } else if (commands[i].equals("3")) {
                System.out.println(
                        editor.get(Integer.parseInt(commands[i + 1])));
            } else {
                editor.undo();
                increment = 1;
            }
            i += increment;
        }

        sc.close();
    }
}

/**
 * CustomStack class represents a stack data structure used for text editing
 * operations.
 * It supports insertion, deletion, retrieval, and undo operations on a sequence
 * of strings.
 */
class CustomStack {

    private Deque<String> editor;
    private Deque<String> temp;

    public CustomStack() {
        this.editor = new ArrayDeque<>();
        this.temp = new ArrayDeque<>();
    }

    /**
     * Inserts the given value into the editor.
     *
     * @param val the value to be inserted
     */
    public void insert(String val) {
        this.editor.addLast(val);
    }

    /**
     * Deletes a specified number of characters from the editor.
     * If the count exceeds the length of the current top editor string, it will
     * delete the entire string.
     *
     * @param count The number of characters to delete.
     */
    public void delete(int count) {
        while (true) {
            int lenTopEditor = this.editor.peekLast().length();
            if (count - lenTopEditor <= 0) {
                break;
            }
            count -= lenTopEditor;
            this.temp.addLast(this.editor.pop());
        }

        // check for deleting the entire string
        if (this.editor.peekLast().length() == count) {
            this.editor.pop();
        } else {
            String last = this.editor.pop();
            this.editor.addLast(last.substring(0, count - 1));
        }

        cleanup();
    }

    /**
     * Retrieves the character at the specified index in the text editor.
     *
     * @param index The index of the character to retrieve.
     * @return The character at the specified index, or '0' if the index is out of
     *         range.
     */
    public char get(int index) {
        for (String text : this.editor) {
            if (index > text.length()) {
                index -= text.length();
                continue;
            }

            return text.charAt(index - 1);
        }
        return '0';
    }

    /**
     * Undoes the last operation performed on the text editor.
     */
    public void undo() {
        this.editor.pop();
    }

    /**
     * Cleans up the temporary stack by transferring its elements back to the editor
     * stack.
     */
    private void cleanup() {
        while (!this.temp.isEmpty()) {
            this.editor.addLast(this.temp.pop());
        }
    }
}
